compiler/rustc_mir_build/src/build/matches/mod.rs

```diff
@@ -461,20 +463,43 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
     /// (by [Builder::lower_match_tree]).
     ///
     /// `outer_source_info` is the SourceInfo for the whole match.
+    /// [Builder::lower_match_tree]).
     fn lower_match_arms(
         &mut self,
         destination: Place<'tcx>,
+        destination_scope: Option<region::Scope>,
         scrutinee_place_builder: PlaceBuilder<'tcx>,
         scrutinee_span: Span,
         arm_candidates: Vec<(&'_ Arm<'tcx>, Candidate<'_, 'tcx>)>,
         outer_source_info: SourceInfo,
         fake_borrow_temps: Vec<(Place<'tcx>, Local, FakeBorrowKind)>,
     ) -> BlockAnd<()> {
+        if arm_candidates.is_empty() {
+            // If there are no arms to schedule drops, then we have to do it
+            // manually.
+            if let Some(scope) = destination_scope {
+                self.schedule_drop(
+                    outer_source_info.span,
+                    scope,
+                    destination.as_local().unwrap(),
+                    DropKind::Value,
+                );
+            }
+            return self.cfg.start_new_block().unit();
+        }
+
+        let mut first_arm = true;
         let arm_end_blocks: Vec<_> = arm_candidates
             .into_iter()
             .map(|(arm, candidate)| {
                 debug!("lowering arm {:?}\ncandidate = {:?}", arm, candidate);
 
+                if first_arm {
+                    first_arm = false;
+                } else if let Some(scope) = destination_scope {
```

Comment:
please add a comment here. this is confusing to me

You're dropping it in the `first_arm` in the call of `this.expr_into_dest(destination, destination_scope, arm_block, arm.body)`. Doesn't each `arm_block` need its own drop of `destination`? :thinking: